ibis/pandas/dispatcher.py

```diff
@@ -0,0 +1,107 @@
+from multipledispatch import Dispatcher
+
+
+class TwoLevelDispatcher(Dispatcher):
+    """An implementation of multipledispatch.Dispatcher that utilizes two
+    levels of dispatching.
+
+    The major change is that this class no longer trigger reorder in
+    dispatch_iter. Because the majority of the slowness is happening
+    in reorder, this implementation makes dispatch_iter faster.
+    Instead, this implementation will trigger reorder in the meta dispatcher
+    and second level dispatcher. Because the number of registered signatures
+    for each dispatcher is much smaller in this implementation (In pandas
+    backend, the number of signatures in one level implementation is
+    O(1000), and the max number of signatures for the meta dispatcher and
+    second level dispatcher is O(100)), the overall dispatch_iter is faster.
+
+    This implementation consist of three Dispatcher instance:
+
+    (1) This dispatcher, or the instance of this class itself. This class
+    inherits Dispatcher to avoid duplicating __call__, cache, ambiguities
+    detection, as well as properties like ordering and funcs.
+
+    (2) First level dispatcher, aka, Meta dispatcher. This is the dispatcher
+    is used to dispatch to the second level dispatcher using the type of the
+    first arg.
+
+    (3) Second level dispatcher. This is the actual dispatcher used for linear
+    searching of matched function given type of args.
+
+    Implementation notes:
+
```

Comment:
can you indicate *when* this goes to 2-level (or it always does)? is there an opt-in that would make this better? e.g. maybe an option to register?

Reply:
or a register_twolevel (better name though)

Reply:
This always goes to 2-level. I think mixing 1-level and 2-level here makes it more complicated and I am not sure it's worth it...(And it's not necessarily more performant). 

The reason that some perf in two-level is slower is because the `reorder` is "amortized" cross all the 2-level dispatchers. 

E.g, the dispatch for `NullIf, Series, float` involves the ordering of the second level dispatcher `NullIf`, which contains > 200 signatures, and therefore takes 0.1 second (might be the highest second level signature number). The dispatch for `NullIf, Series, float` in the one-level implementation doesn't trigger any reordering (it has been ordered already), so it's fast. But this doesn't necessarily mean the dispatching for `NullIf, Series, float` is worst in two-level impl, it's just that the time to order signatures for `NullIf` is taken somewhere else in the one-level impl.

Reply:
kk sounds good, yeah just wanted to you to think thru the complexities. agree anything more is too complex.